lib/agent.js

```diff
@@ -450,18 +430,58 @@ Agent.prototype.startAggregators = function startAggregators() {
   }
 }
 
-Agent.prototype.onConnect = function onConnect() {
-  this.metrics.reconfigure(this.config)
-  this.errors.reconfigure(this.config)
-  this.traces.reconfigure(this.config)
-  this.queries.reconfigure(this.config)
-  this.spanEventAggregator.reconfigure(this.config)
-  this.transactionEventAggregator.reconfigure(this.config)
-  this.customEventAggregator.reconfigure(this.config)
+/**
+ * Completes any final setup upon full connection to New Relic
+ * servers and sets the agent state to 'started'.
+ */
+Agent.prototype.onConnect = function onConnect(shouldImmediatelyHarvest, callback) {
+  this._reconfigureAggregators(this.config)
 
   if (this.config.certificates && this.config.certificates.length > 0) {
     this.metrics.getOrCreateMetric(NAMES.FEATURES.CERTIFICATES).incrementCallCount()
   }
+
+  this._scheduleHarvests(shouldImmediatelyHarvest, callback)
+
+  this.setState('started')
+}
+
+Agent.prototype._reconfigureAggregators = function _reconfigureAggregators(config) {
+  this.metrics.reconfigure(config)
+  this.errors.reconfigure(config)
+  this.traces.reconfigure(config)
+  this.queries.reconfigure(config)
+  this.spanEventAggregator.reconfigure(config)
+  this.transactionEventAggregator.reconfigure(config)
+  this.customEventAggregator.reconfigure(config)
+}
+
+Agent.prototype._scheduleHarvests = function _scheduleHarvests(shouldImmediatelyHarvest, callback) {
+  if (!shouldImmediatelyHarvest) {
+    this.startAggregators()
+    setImmediate(callback)
```

Comment:
Not obvious from refactor but I introduced an async hop via `setImmediate` for the case where aggregators are simply started for consistency with the "immediate" harvest case that gets scheduled via `setTimeout`.